src/main/java/me/zeroeightsix/kami/util/graphics/TextComponent.kt

```diff
@@ -0,0 +1,188 @@
+             skipEmptyLine: Boolean = true,
+             horizontalAlign: HAlign = HAlign.LEFT,
+             verticalAlign: VAlign = VAlign.TOP
+    ) {
+        if (isEmpty()) return
+        glPushMatrix()
+        glTranslated(round(pos.x), round(pos.y) + 1.0, 0.0) // Rounding it to int so stupid Minecraftia doesn't fucked up
+        glScalef(scale, scale, 1f)
+        if (verticalAlign != VAlign.TOP) {
+            var height = getHeight(lineSpace)
+            if (verticalAlign == VAlign.CENTER) height /= 2
+            glTranslatef(0f, -height.toFloat(), 0f)
+        }
+        for (line in textLines) {
+            if (skipEmptyLine && (line == null || line.isEmpty())) continue
+            line?.drawLine(fontRenderer, drawShadow, horizontalAlign)
+            glTranslatef(0f, (fontRenderer.FONT_HEIGHT + lineSpace).toFloat(), 0f)
+        }
+        glPopMatrix()
+    }
+
+    fun isEmpty() = textLines.firstOrNull { it?.isEmpty() == false } == null
+
+    fun getWidth() = textLines.map { it?.getWidth(fontRenderer) ?: 0 }.max() ?: 0
+
+    fun getHeight(lineSpace: Int, skipEmptyLines: Boolean = false) = fontRenderer.FONT_HEIGHT * getLines(skipEmptyLines) + lineSpace * (getLines(skipEmptyLines) - 1)
+
+    fun getLines(skipEmptyLines: Boolean = true) = textLines.count { !skipEmptyLines || (it != null && !it.isEmpty()) }
+
+    override fun toString() = textLines.joinToString(separator = "\n")
+
+    private class TextLine(val separator: String) {
+        private val textElementList = ArrayList<TextElement>()
+
+        fun isEmpty() = textElementList.size == 0
+
+        fun add(textElement: TextElement) {
+            textElementList.add(textElement)
+        }
+
+        fun drawLine(fontRenderer: FontRenderer, drawShadow: Boolean = false, horizontalAlign: HAlign) {
+            glPushMatrix()
+            if (horizontalAlign != HAlign.LEFT) {
+                var width = getWidth(fontRenderer)
+                if (horizontalAlign == HAlign.CENTER) width /= 2
+                glTranslatef(-width.toFloat(), 0f, 0f)
+            }
+            for (textElement in textElementList) {
+                val width = fontRenderer.getStringWidth(textElement.text + separator)
+                fontRenderer.drawString(textElement.text, 0f, 0f, textElement.color, drawShadow)
+                glTranslatef(width.toFloat(), 0f, 0f)
+            }
+            glPopMatrix()
+        }
+
+        fun getWidth(fontRenderer: FontRenderer): Int {
+            return fontRenderer.getStringWidth(toString())
+        }
+
+        override fun toString() = textElementList.joinToString(separator = separator)
+    }
+
+    class TextElement(textIn: String, val color: Int = 0xFFFFFF, val style: TextStyle = TextStyle.REGULAR) {
+        val text = "§r${style.code}$textIn"
+
+        override fun toString(): String {
+            return text
+        }
+    }
+
+    @Suppress("UNUSED")
+    enum class TextStyle(val code: String) {
```

Comment:
what about colors?

Reply:
We use hex color instead or Minecraft text formatting color code.